### ISSUE TITLE
Variant matching conditions allow malformed code

### ISSUE THREAD

**Issue by fixplz** [facebook/flow#1348]
`f1` and `f2` should be rejected here but aren't:

```js
type Opts = {bzt:{}}

type Variant = { type: 'a', opts_a: Opts } | { type: 'b', opts_b: Opts }

function f0 (val : Variant) {
  if(val.type === 'a') {
    g(val.opts_a)
  }
}
function f1 (val : Variant) {
  if(val.wat.type === 'a') {
    g(val.wat.wat)
  }
}
function f2 (val : {}) {
  if(val.wat.type === 'a') {
    g(val.wat.wat)
  }
}

function g (opts : Opts) {
}
```

Looks like `val.wat.type === 'a'` makes it believe that `val.wat : any`.

Using Flow 0.20.1.

**Comment by samwgoldman** [comment]
Where were you a few weeks ago? This one is *also* fixed in the next release!

Thank you for reporting these issues. The repro cases are great. Please keep reporting them; inevitably we'll hit one that isn't fixed in master. :)

**Comment by fixplz** [comment]
Ohh, makes sense. Guess I'll use repo versions to test next.

Should I close this?

**Comment by samwgoldman** [comment]
We released 0.21.0 which should have the fix for this. Can you confirm the issue is resolved? If it is, then yeah, please close this. Otherwise let's dig deeper.

**Comment by fixplz** [comment]
Yep it is fixed.
